core/native/src/main/scala/fs2/hash.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2013 Functional Streams for Scala
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2
+
+import cats.effect.kernel.Sync
+
+import scala.scalanative.unsafe._
+import scala.scalanative.unsigned._
+
+/** Provides various cryptographic hashes as pipes. Requires OpenSSL. */
+object hash {
+  import openssl._
+
+  /** Computes an MD2 digest. */
+  def md2[F[_]: Sync]: Pipe[F, Byte, Byte] = digest(EVP_get_digestbyname(c"MD2"))
+
+  /** Computes an MD5 digest. */
+  def md5[F[_]: Sync]: Pipe[F, Byte, Byte] = digest(EVP_get_digestbyname(c"MD5"))
+
+  /** Computes a SHA-1 digest. */
+  def sha1[F[_]: Sync]: Pipe[F, Byte, Byte] =
+    digest(EVP_get_digestbyname(c"SHA1"))
+
+  /** Computes a SHA-256 digest. */
+  def sha256[F[_]: Sync]: Pipe[F, Byte, Byte] =
+    digest(EVP_get_digestbyname(c"SHA256"))
+
+  /** Computes a SHA-384 digest. */
+  def sha384[F[_]: Sync]: Pipe[F, Byte, Byte] =
+    digest(EVP_get_digestbyname(c"SHA384"))
+
+  /** Computes a SHA-512 digest. */
+  def sha512[F[_]: Sync]: Pipe[F, Byte, Byte] =
+    digest(EVP_get_digestbyname(c"SHA512"))
+
+  /** Computes the digest of the source stream, emitting the digest as a chunk
+    * after completion of the source stream.
+    */
+  private[this] def digest[F[_]](digest: => Ptr[EVP_MD])(implicit F: Sync[F]): Pipe[F, Byte, Byte] =
```

Comment:
So I caved and added a `Sync` constraint here, unlike the JVM/JS implementations. Since we are allocating and `free`ing native memory I didn't see a way around it.

This might end up being annoying in practice, especially for cross-builds. Maybe a capability typeclass `Hash[F]` would be a better design.